src/sample.rs

```diff
@@ -27,173 +29,258 @@ use std::str::FromStr;
-							kinds.push(k);
-						}
-					}
-					let since = Timestamp::from_str(since_raw.unwrap()).unwrap();
-					let until = Timestamp::from_str(until_raw.unwrap()).unwrap();
-					let filter = Filter::new().kinds(kinds).since(since).until(until);
-					let client_message = ClientMessage::new_req(id, vec![filter]);
-					let serialized_message = client_message.as_json();
-					tx.unbounded_send(Message::text(serialized_message)).unwrap();
-				},
-				"closesubscription" => {
-					let subscriptionid = words.next();
-					if subscriptionid.is_none() {
-						println!("ERROR: closesubscription has 1 required argument: `closesubscription subscriptionid");
-					}
-					let id = SubscriptionId::new(subscriptionid.unwrap());
-					let client_message = ClientMessage::close(id);
-					let serialized_message = client_message.as_json();
-					tx.unbounded_send(Message::text(serialized_message)).unwrap();
-				},
-				"shutdown" => {
-					tx.unbounded_send(Message::Close(None)).unwrap();
-					tx.close_channel();
-					println!("Civkit sample exiting...");
-					process::exit(0x0100);
-				},
-				_ => { println!("Unknown command !"); },
-			}
-		}
-	}
+    println!("Civkit sample startup successful. Enter \"help\" to view available commands");
+
+    loop {
+        print!("> ");
+        io::stdout().flush().unwrap();
+        let mut line = String::new();
+        if let Err(e) = io::stdin().read_line(&mut line) {
+            break println!("ERROR {}", e);
+        }
+
+        if line.len() == 0 {
+            continue;
+        }
+
+        match respond(&line, &tx, &client_keys) {
+            Ok(quit) => {
+                if quit {
+                    process::exit(0x0100);
+                }
+            }
+            Err(err) => {
+                write!(std::io::stdout(), "{err}").expect("error: Failed to write to stdout");
+                std::io::stdout()
+                    .flush()
+                    .expect("error: Failed to flush stdout");
+            }
+        }
+    }
+}
+
+fn cli() -> Command {
+    // strip out usage
+    const PARSER_TEMPLATE: &str = "\
+        {all-args}
+    ";
+    // strip out name/version
+    const APPLET_TEMPLATE: &str = "\
+        {about-with-newline}\n\
+        {usage-heading}\n    {usage}\n\
+        \n\
+        {all-args}{after-help}\
+    ";
+
+    Command::new("Nostr Client REPL")
+        .multicall(true)
+        .arg_required_else_help(true)
+        .subcommand_required(true)
+        .help_template(PARSER_TEMPLATE)
+        .subcommand(
+            Command::new("sendtextnote")
+                .args([Arg::new("content").help("the content of the text note").required(true)])
+                .help_template(APPLET_TEMPLATE)
+                .about("Send a text note to the relay"),
+        )
+        .subcommand(
+            Command::new("setmetadata")
+                .args([
+					Arg::new("username").help("The client's username").required(true),
```

Comment:
could be aligned though nevermind

we could add the same coding conventions than LDK’s one here: https://github.com/lightningdevkit/rust-lightning/blob/main/CONTRIBUTING.md#coding-conventions to be honest for now i’m quite shrug about strong rust coding conventions